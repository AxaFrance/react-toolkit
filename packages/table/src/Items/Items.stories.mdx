import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { ArgsTable } from '@storybook/addon-docs';
import Items from './Items';

<Meta
  title="Components/Table/Paging/Items"
  component={Items}
  argTypes={{
    onChange: { action: 'changed' },
    items: {
      control: {
        type: 'array',
        options: [5, 10, 25, 50, 100],
      },
    },
    numberItems: {
      control: {
        type: 'select',
        options: [5, 10, 25, 50, 100],
      },
    },
  }}
/>

# Table paging items

export const Template = (args) => <Items {...args} />;

The Items component allows you to change how many items a table component will show.  
The `numberItems` property carry that number, as the `items` property allows you to change the list of elements to display in the table.

You will have to install the `select` component in order to have the correct style for this component.

```shell
npm i @axa-fr/react-toolkit-form-input-select
npm i @axa-fr/react-toolkit-form-core
```

The associed stylesheet are the ones from the `select` component of the toolkit.

```javascript
import '@axa-fr/react-toolkit-form-input-select/dist/af-select.css';
import '@axa-fr/react-toolkit-form-core/dist/af-form.css';
```

<Canvas>
  <Story
    name="Items"
    args={{
      items: [5, 10, 25, 50, 100],
      numberItems: 5,
    }}>
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Items" />
