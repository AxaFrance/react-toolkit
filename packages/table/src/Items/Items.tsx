import React from 'react';
import {
  WithClassModifierOptions,
  InputManager,
  withClassDefault,
  withClassModifier,
  WithOnChangeEvent,
  OnChangeCustomEvent,
  compose,
} from '@axa-fr/react-toolkit-core';

export type ItemsComponentProps = Pick<
  React.HTMLProps<HTMLAnchorElement>,
  'className'
> &
  WithOnChangeEvent<OnChangeCustomEvent> & {
    id: string;
    numberItems?: number;
    items?: number[];
    displayLabel?: React.ReactNode;
    elementsLabel?: React.ReactNode;
  };
const DEFAULT_CLASSNAME = 'af-paging__form';

const defaultProps: Partial<ItemsComponentProps> = {
  numberItems: 10,
  items: [5, 10, 25, 50, 100],
  id: null,
  displayLabel: 'Afficher',
  elementsLabel: 'éléments',
};

class Items extends React.PureComponent<ItemsComponentProps> {
  constructor(props: ItemsComponentProps) {
    super(props);
    this.onChange = this.onChange.bind(this);
  }

  onChange(event: React.ChangeEvent<HTMLSelectElement>) {
    event.preventDefault();
    const { onChange } = this.props;
    onChange({
      value: Number(event.target.value),
    });
  }

  render() {
    const {
      className,
      displayLabel,
      elementsLabel,
      id,
      items,
      numberItems,
    } = this.props;

    const content = items.map(item => (
      <option key={item} value={item}>
        {item}
      </option>
    ));

    const defaultIdName = InputManager.getInputId(id);
    return (
      <div className="af-paging__limit">
        <form className={className}>
          <div className="af-form__group">
            <div className="col col-sm-2 col-md-2 col-lg-2 col-xl-2">
              <label className="af-form__group-label" htmlFor={defaultIdName}>
                {displayLabel}
              </label>
            </div>
            <div className="col col-sm-10 col-md-10 col-lg-10 col-xl-10">
              <div className="af-form__select">
                <div className="af-form__select-container">
                  <select
                    id={defaultIdName}
                    className="af-form__input-select"
                    value={numberItems}
                    onChange={this.onChange}>
                    {content}
                  </select>
                  <span className="glyphicon glyphicon-menu-down" />
                </div>
                <span className="af-form__input-cmplt">{elementsLabel}</span>
              </div>
            </div>
          </div>
        </form>
      </div>
    );
  }
}
export type ItemsProps = ItemsComponentProps & WithClassModifierOptions;

const enhance = compose<ItemsComponentProps, ItemsProps>(
  withClassDefault(DEFAULT_CLASSNAME),
  withClassModifier
);
const Enhance = enhance(Items);
Enhance.defaultProps = defaultProps;
export default Enhance;
