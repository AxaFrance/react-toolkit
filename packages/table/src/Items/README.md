## Items

### Installation

```sh
npm i @axa-fr/react-toolkit-table
```

### Import

```javascript
import { Items } from '@axa-fr/react-toolkit-table';
```
