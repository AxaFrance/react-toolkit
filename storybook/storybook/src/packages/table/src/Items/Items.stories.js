import React from 'react';
import { action } from '@storybook/addon-actions';
import { text, number } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { Items } from '@axa-fr/react-toolkit-table';
import readme from '@axa-fr/react-toolkit-table/src/Items/README.md';

const ItemsStory = () => (
  <Items
    classModifier={text('classModifier', '')}
    className={text('className', '')}
    displayLabel={text('displayLabel', 'Show')}
    elementsLabel={text('elementsLabel', 'elements')}
    items={[5, 10, 25, 50, 100]}
    onChange={action('onChange')}
    numberItems={number('numberItems', 20)}
  />
);

const stories = storiesOf('Table/Items', module);

stories.addParameters({
  readme: {
    sidebar: readme,
  },
  options: {},
});

stories.add('Default', ItemsStory);
