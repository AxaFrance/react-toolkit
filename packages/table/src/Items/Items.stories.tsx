import React from 'react';
import { Story } from '@storybook/react';
import Items, { ItemsProps } from './Items';
import readme from './README.md';

export default {
  title: 'Table/Items',
  component: Items,
  parameters: {
    readme: {
      sidebar: readme,
    },
    options: {},
  },
  argTypes: {
    onChange: { action: 'changed' },
    items: {
      control: {
        type: 'array',
        options: [5, 10, 25, 50, 100],
      },
    },
    numberItems: {
      control: {
        type: 'select',
        options: [5, 10, 25, 50, 100],
      },
    },
  },
};

const Template: Story<ItemsProps> = args => <Items {...args} />;

export const Default = Template.bind({});
Default.args = {
  classModifier: '',
  className: '',
  displayLabel: 'Show',
  elementsLabel: 'elements',
};
