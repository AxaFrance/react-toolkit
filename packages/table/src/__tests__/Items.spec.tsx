import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Items from '../Items/Items';

describe('Table.Items', () => {
  it('renders correctly with default values', async () => {
    const handleOnChangeMock = jest.fn();
    handleOnChangeMock.mockImplementationOnce(() => 'My mock is called');

    const { asFragment } = render(
      <Items
        numberItems={10}
        onChange={handleOnChangeMock}
        items={[5, 10, 25, 50, 100]}
        id="itemid"
      />
    );

    await screen.findByText('Afficher');

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with custom values', async () => {
    const handleOnChangeMock = jest.fn().mockReturnValue('My mock is called');

    const { getByRole, getByLabelText, getByText } = render(
      <Items
        numberItems={10}
        onChange={handleOnChangeMock}
        items={[5, 10, 25, 50, 100]}
        id="itemid"
        displayLabel="Show"
        elementsLabel="elements"
      />
    );
    await screen.findByText('Show');

    expect(getByLabelText('Show')).toBeInTheDocument();
    expect(getByText(/elements/)).toBeInTheDocument();

    await userEvent.selectOptions(getByRole('combobox'), ['50']);
    expect(handleOnChangeMock).toHaveBeenCalled();
  });
});
