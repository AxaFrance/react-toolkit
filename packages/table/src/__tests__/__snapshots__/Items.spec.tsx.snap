// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.Items renders correctly with default values 1`] = `
<DocumentFragment>
  <div
    class="af-paging__limit"
  >
    <form
      class="af-paging__form"
    >
      <div
        class="af-form__group"
      >
        <div
          class="col col-sm-2 col-md-2 col-lg-2 col-xl-2"
        >
          <label
            class="af-form__group-label"
            for="itemid"
          >
            Afficher
          </label>
        </div>
        <div
          class="col col-sm-10 col-md-10 col-lg-10 col-xl-10"
        >
          <div
            class="af-form__select"
          >
            <div
              class="af-form__select-container"
            >
              <select
                class="af-form__input-select"
                id="itemid"
              >
                <option
                  value="5"
                >
                  5
                </option>
                <option
                  value="10"
                >
                  10
                </option>
                <option
                  value="25"
                >
                  25
                </option>
                <option
                  value="50"
                >
                  50
                </option>
                <option
                  value="100"
                >
                  100
                </option>
              </select>
              <span
                class="glyphicon glyphicon-menu-down"
              />
            </div>
            <span
              class="af-form__input-cmplt"
            >
              éléments
            </span>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
